### ISSUE TITLE
best interface for gf_abline() and friends?

### ISSUE THREAD

**Issue by rpruim** [ProjectMOSAIC/ggformula#33]
I'm trying to decide how/whether we want to support plotting multiple lines in one call of `gf_abline()`, `gf_hline()`, and `gf_vline()`.  These functions don't fit our usual pattern exactly because the formula isn't really used, at least not it the most common case: adding a single line.

It might be worth writing a separate factory function for these and tuning it for their use.  (If you look at `geom_abline()` you will see that it does some preprocessing to create a data frame used for the plot and that the data frame passed in as `data` is typically ignored.  We might want to do something similar.

**Comment by rpruim** [comment]
This is a little tricky, even in `ggplot2`, especially when faceting.  But I think I have a workable solution.  There are several options for using these functions.

###  layer-specific data

`data` provided to the layer can contain slope and/or intercept values for the lines.  Faux-mapping will use these values as the actual slope and/or intercept for a line.  (So they aren't really mapped in the usual sense of the work, but a `~` indicates that they come from the data rather than being set to a constat.) 

```r
mtcars2 <- df_stats( wt ~ cyl, data = mtcars)
gf_point(wt ~ hp, size = ~wt, color = ~cyl, data = mtcars) %>%
  gf_abline(slope = 0, intercept = ~median, color = ~cyl, data = mtcars2)
gf_point(wt ~ hp, size = ~wt, color = ~cyl, data = mtcars) %>%
  gf_hline(slope = 0, yintercept = ~median, color = ~cyl, data = mtcars2)
```

![image](https://user-images.githubusercontent.com/722231/27591704-7cf47e6a-5b20-11e7-9ff1-8688baf41886.png)


### data = NA

I'm currently making this the default for these functions.  It does the trick that `geom_abline()` does and creates a data frame from the attributes that are set or mapped and uses that data frame to construct the plot.  For the slope and intercept, setting and mapping are the same.  For other attributes (like color), they are different.  Use setting to directly set the colors and mapping to map them to values of a variable.

```r
gf_hline(color = "orange", yintercept = 20, data = NA) %>%
   gf_vline(color = ~c("4", "6", "8"), xintercept = c(80, 120, 250), data = NA) %>%
   gf_point(mpg ~ hp, size = ~wt, color = ~ factor(cyl), data = mtcars, alpha = 0.3)
```

![image](https://user-images.githubusercontent.com/722231/27590965-503fb27e-5b1e-11e7-9753-9bddb0711a3a.png)

The `data = NA` option could potentially be used with other gf functions as well to create simple data sets on the fly.  But to make that really work, I'd have to write things to ignore the formula when `data = NA`.  I haven't decided if there is really a use case here.


### data = NULL

This is the default and means "use the data from `ggplot()`, i.e. the primary data for the plot.  This seems less useful than `data = NA`, but I guess you can do silly stuff like

```r
gf_point(slope ~ 1, data = D) %>% gf_abline(slope = ~ slope, intercept = ~intercept, data = NULL)
```

Basically, I'm imagining two use cases

  1. supply data
  2. default (`data = NA` here, `data = NULL` most other places)

Distinguishing between `NA` and `NULL` lets me generate these with a common wrapper.
